pkg/solana/txm/pendingtx.go

```diff
@@ -138,28 +196,270 @@ func (c *pendingTxContext) ListAll() []solana.Signature {
+	c.finalizedErroredTxs[id] = tx
+	return id, nil
+}
+
+func (c *pendingTxContext) OnError(sig solana.Signature, retentionTimeout time.Duration, _ int) string {
+	// check if already cancelled
+	c.lock.RLock()
+	id, sigExists := c.sigToID[sig]
+	if !sigExists {
+		c.lock.RUnlock()
+		return ""
+	}
+	// transaction can transition from any non-finalized state
+	var broadcastedExists, confirmedExists bool
+	_, broadcastedExists = c.broadcastedTxs[id]
+	_, confirmedExists = c.confirmedTxs[id]
+	// transcation does not exist in any tx maps
+	if !broadcastedExists && !confirmedExists {
+		c.lock.RUnlock()
+		return ""
+	}
+	c.lock.RUnlock()
+
+	// upgrade to write lock if sig exists
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if _, exists := c.sigToID[sig]; !exists {
+		return ""
+	}
+	var tx, tempTx pendingTx
+	if tempTx, broadcastedExists = c.broadcastedTxs[id]; broadcastedExists {
+		tx = tempTx
+	}
+	if tempTx, confirmedExists = c.confirmedTxs[id]; confirmedExists {
+		tx = tempTx
+	}
+	// transcation does not exist in any non-finalized maps
+	if !broadcastedExists && !confirmedExists {
+		c.lock.RUnlock()
+		return ""
+	}
+	// call cancel func + remove from map
+	if cancel, exists := c.cancelBy[id]; exists {
+		cancel() // cancel context
+		delete(c.cancelBy, id)
+	}
+	// delete from broadcasted map, if exists
+	delete(c.broadcastedTxs, id)
+	// delete from confirmed map, if exists
+	delete(c.confirmedTxs, id)
+	// remove all related signatures from the sigToID map to skip picking up this tx in the confirmation logic
+	for _, s := range tx.signatures {
+		delete(c.sigToID, s)
+	}
+	// if retention duration is set to 0, delete transaction from storage
+	// otherwise, move to finalized map
+	if retentionTimeout == 0 {
+		return id
+	}
+	// set current timestamp when transitioning to finalized map
+	// later used for transaction retention
+	tx.retentionTs = time.Now().Add(retentionTimeout)
+	// update tx state to Errored
+	tx.state = Errored
+	// move transaction from broadcasted to error map
+	c.finalizedErroredTxs[id] = tx
+	return id
+}
+
+func (c *pendingTxContext) GetTxState(id string) (TxState, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	if tx, exists := c.broadcastedTxs[id]; exists {
+		return tx.state, nil
+	}
+	if tx, exists := c.confirmedTxs[id]; exists {
+		return tx.state, nil
+	}
+	if tx, exists := c.finalizedErroredTxs[id]; exists {
+		return tx.state, nil
+	}
+	return NotFound, fmt.Errorf("failed to find transaction for id: %s", id)
+}
+
+// TrimFinalizedErroredTxs deletes transactions from the finalized/errored map and the allTxs map after the retention period has passed
+func (c *pendingTxContext) TrimFinalizedErroredTxs() {
+	c.lock.RLock()
+	var expiredIDs []string
+	for id, tx := range c.finalizedErroredTxs {
+		if time.Now().After(tx.retentionTs) {
+			expiredIDs = append(expiredIDs, id)
+		}
+	}
+	c.lock.RUnlock()
+
+	// upgrade to write lock
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	for _, id := range expiredIDs {
```

Comment:
- Should we worry about this being unbounded if the number of txes is too large? 
- Should we add an upper bound? 
- Should we add a note to the TxReapInterval config to make this visible?

Reply:
Great! My concern was that if the number of transactions we need to delete is large and frequent, we may be locking a lot and will affect the actual processing of the transactions

But thinking it again, better not to over engineer too soon and check later if this is a thing. In that case we could:
- Limit number of deletions on each pass as a simple approach
- Analyze if having different locks makes the processing independent of the reaping